### ISSUE TITLE
Specify theme in appsetting.json

### ISSUE THREAD

**Issue by jp7677** [serilog/serilog-sinks-console#16]
Hi there,

Thanks a lot for creating a console logger that uses ansi codes!. One question though, is there a way to specify the theme using appsettings.json?
When using 
```
"WriteTo": [
      { "Name": "Console",
        "Args": {
          "theme": "AnsiConsoleTheme.Literate",
          "outputTemplate": "{Timestamp:yyyy-MM-dd HH:mm:ss zzz}|{Level}|{ThreadId}|{RequestId}|{SourceContext}{NewLine}    {Message}{NewLine}{Exception}"
        } 
      }
    ],

```
I'm ending with a

```
Unhandled Exception: System.InvalidCastException: Invalid cast from 'System.String' to 'Serilog.Sinks.SystemConsole.Themes.ConsoleTheme'.
   at System.Convert.DefaultToType(IConvertible value, Type targetType, IFormatProvider provider)
   at Serilog.Settings.Configuration.StringArgumentValue.ConvertTo(Type toType)
   at Serilog.Settings.Configuration.ConfigurationReader.<>c.<CallConfigurationMethods>b__14_4(<>f__AnonymousType6`2 <>h__TransparentIdentifier0)
   at System.Linq.Enumerable.<>c__DisplayClass151_0`3.<CombineSelectors>b__0(TSource x)
   at System.Linq.Enumerable.SelectListPartitionIterator`2.ToList()
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at Serilog.Settings.Configuration.ConfigurationReader.CallConfigurationMethods(ILookup`2 methods, IList`1 configurationMethods, Object receiver)
   at Serilog.Configuration.LoggerSettingsConfiguration.Settings(ILoggerSettings settings)
```
Best regards,
Jens

**Comment by nblumhardt** [comment]
Thanks for the report, Jens! Agreed it would be nice here if we had some support - I think this'd have to happen upstream in Serilog's key-value settings system, or in https://github.com/serilog/serilog-extensions-configuration. A bit short on time to investigate, but will leave this open as a flag to anyone who might be able to get a design together for it. Cheers!

**Comment by tsimbalar** [comment]
One option might be to defined an `enum`, say `PredefinedConsoleTheme`with short names for all the predefined themes, like: 
```csharp
public enum PredefinedConsoleTheme
{
    None,
    Default, // aka Literate
    Grayscale,
    AnsiLiterate,
    AnsiGrayscale,
    AnsiVsCode
}
```

... and them adding an overload of https://github.com/serilog/serilog-sinks-console/blob/4a95053872f32d3f96d0cca62a129dc17b6b0fee/src/Serilog.Sinks.Console/ConsoleLoggerConfigurationExtensions.cs#L48-L56

that would accept a `PredefinedConsoleTheme` instead of a `ConsoleTheme`

```csharp
public static LoggerConfiguration Console(
    this LoggerSinkConfiguration sinkConfiguration,
    LogEventLevel restrictedToMinimumLevel = LevelAlias.Minimum,
    string outputTemplate = DefaultConsoleOutputTemplate,
    IFormatProvider formatProvider = null,
    LoggingLevelSwitch levelSwitch = null,
    LogEventLevel? standardErrorFromLevel = null,
    PredefinedConsoleTheme predefinedTheme = PredefinedConsoleTheme.Default)
```

Using an `enum`means that *Serilog.Settings.AppSettings* and *Serilog.Settings.Configuration* would already know how to handle those values ...

Would that sound reasonable ? I'm not sure if that would introduce an ambiguity in the *extension methods* though ....  

(*predefined* or *builtin* or any other name that sounds better :) )

**Comment by nblumhardt** [comment]
This was my thought too, but I'm put off by the additional parameter needing to be in the first argument position with no default value in order not to upset C# overload resolution (this is why the `ITextFormatter` arguments in other places have that position/form). Either that, or we'd need to include both the typed theme and enum theme arguments within the one method signature.

Still, although it'd be nice to make static readonly property names work with the various settings readers, unlike type names there's no canonical way to represent these (the assembly name really needs to be able to be included in order for it to work in all scenarios).

It's a shame we can't somehow define the enum as you suggest @tsimbalar , but instead of adding a parameter, define some kind of conversion from enum -> static member......... Not that I've ever liked .NET's built-in custom type converter infrastructure.....

**Comment by snoozbuster** [comment]
Given that the `Serilog.Settings.Configuration` library already supports using a fully qualified type string as a parameter and can create such a type given that all its constructor arguments use default parameters, one solutions is to create subclasses of `ConsoleTheme` (or its subclasses as appropriate) that each have parameterless constructors, and then specify them that way instead of via static members (the static members could even resolve to the subclass variants). However, `Serilog.Settings.Configuration` has an artificial restriction that the destination type of such a conversion be an interface type, and that would have to be relaxed or removed.

You could even make these subclasses internal-only, as `EmptyConsoleTheme` is, but set an `[assembly:InternalsVisibleTo("Serilog.Settings.Configuration")]` attribute in the AssemblyInfo to make the classes invisible to anyone using in-code definition. This would, of course, require explicit documentation to make people aware of these types, and creating subclasses for each predefined theme is obnoxious, but it would work pretty well.

Another solution would be a `string` -> `ConsoleTheme` explicit cast operator on the abstract class, which would take the string, convert it to the `PredefinedConsoleTheme` enum, and then return the appropriate console theme from the appropriate static property. Less work in the implementation, maybe would make more sense to anyone reading the appsettings, but would be pretty weird to anyone comparing the C# method with the appsettings value.

**Comment by snoozbuster** [comment]
A third solution would be to extend the type string syntax with something like `Serilog.Sinks.SystemConsole.Themes.ConsoleTheme::None, Serilog.Sinks.Console, Version=3.1.0.0, Culture=neutral, PublicKeyToken=24c2f752a8e58a10`, and have the processing in `Serilog.Settings.Configuration` look for the `::Member` syntax, strip it out, resolve the type, and then look for a static property on it by the name `Member` and use that. 

Of the three, the first I proposed is (in my opinion) the most intuitive, but most work. The other two are less work but each less intuitive and (again, in my opinion) hacky.

**Comment by tsimbalar** [comment]
@snoozbuster some pretty good ideas ! I agree that the first option you mention would be the most straightforward and be the least surprising for the end user. The last option sounds nice, but I don't know if it's worth adding new *syntax* to the key-value settings format for just one specific use-case.

From what I understand, your first option means : 

By combining 
- serilog/serilog-settings-configuration#71 (to allow the `Serilog.Settings.*` to accept a fully qualified concrete class name when a sink argument's type is an `abstract` class)
- making each predefined theme a *concrete* subclass of `abstract` class `ConsoleTheme`

We could possibly also express the setting as : 

```xml
<appsettings>
    ...
    <add key="serilog:write-to:Console.theme" value="Serilog.Sinks.Console.Themes.AnsiGrayscaleTheme, Serilog.Sinks.Console, Version=3.1.0.0, Culture=neutral, PublicKeyToken=24c2f752a8e58a10" />
    ...
</appsettings>
```

or

```json
"WriteTo": [
      { "Name": "Console",
        "Args": {
          "theme": "Serilog.Sinks.Console.Themes.AnsiGrayscaleTheme, Serilog.Sinks.Console, Version=3.1.0.0, Culture=neutral, PublicKeyToken=24c2f752a8e58a10",
          "outputTemplate": "{Timestamp:yyyy-MM-dd HH:mm:ss zzz}|{Level}|{ThreadId}|{RequestId}|{SourceContext}{NewLine}    {Message}{NewLine}{Exception}"
        } 
      }
    ],
```

and that would also allow the user to inject *custom* implementations of `ConsoleTheme`.

It's not ideal nor particularly elegant, but at least it makes it possible :) 

Also maybe there would be a way to relax the *fully-qualified* type names when the type is in the same assembly as the `abstract` class, or something like that ?

**Comment by tsimbalar** [comment]
Hi,

The support for setting values such as `Serilog.Sinks.SystemConsole.Themes.ConsoleTheme::None, Serilog.Sinks.Console` has landed in serilog/serilog#1064 serilog key-value settings format (and therefore in `Serilog.Settings.AppSettings`) and it is available from Serilog v2.6.0+. 

There is an *up-for-grabs* issue serilog/serilog-settings-configuration#73
 to implement a similar behavior in *Serilog.Settings.Configuration* if someone feels like taking a shot ;)
